src/server/route/routes/user.ts

```diff
@@ -0,0 +1,110 @@
+import { router } from '../router';
+import { Request, Response } from 'express';
+import { db } from '../../dbConfigs';
+import { Login, Registration } from '../../../../types';
+import bcrypt from 'bcrypt';
+import * as jwt from 'jsonwebtoken';
+import dotenv from 'dotenv';
+dotenv.config();
+
+router.route('/user/register').post(async (req: Request, res: Response) => {
+    const user: Registration = req.body;
+
+    if (!user || !user.username || !user.password || !user.email) {
+        res.status(403).json({ message: 'Missing parameters' });
+        return;
+    }
+
+    const saltRounds = 10;
+    const hash = await bcrypt.hash(user.password, saltRounds);
+    const q = await db.query(
+        'SELECT username FROM users WHERE username=$1 OR email=$2',
+        [user.username, user.email]
+    );
+    if (q.rowCount == 0) {
+        await db.query(
+            'INSERT INTO users (username, password, email) VALUES ($1, $2, $3)',
+            [user.username, hash, user.email]
+        );
+        res.status(200).json();
+    } else {
+        res.status(403).json({
+            message: 'Username or email already registered',
+        });
+    }
+});
+
+router.route('/user/login').post(async (req: Request, res: Response) => {
+    const body: Login = req.body;
+    let user = null;
+
+    if (!body.password || (!body.email && !body.username)) {
+        res.status(403).json({ message: 'Missing parameters' }).end();
+        return;
+    }
+
+    //User tries to login with email
+    if (body.email && body.email.includes('@')) {
+        const { rows, rowCount } = await db.query(
+            'SELECT * FROM users WHERE email=$1',
+            [body.email]
+        );
+
+        if (rowCount == 0) {
+            res.status(401).json({ message: 'Wrong email or password' }).end();
+            return;
+        } else if (rowCount >= 2) {
+            res.status(500).json({ message: 'More than one user found' }).end();
+            return;
+        }
+
+        user = rows[0];
+    }
+
+    //User tries to login with a username
+    if (body.username) {
+        const { rows, rowCount } = await db.query(
+            'SELECT * FROM users WHERE username=$1',
+            [body.username]
+        );
+
+        if (rowCount == 0) {
+            res.status(401)
+                .json({ message: 'Wrong username or password' })
+                .end();
+            return;
+        } else if (rowCount >= 2) {
+            res.status(500).json({ message: 'More than one user found' }).end();
+            return;
+        }
+
+        user = rows[0];
+    }
+
+    const passCorrect = await bcrypt.compare(body.password, user.password);
+
+    if (!passCorrect) {
+        res.status(401).json({ message: 'Wrong email or password' });
+        return;
+    }
+
+    // Cannot give a type. Might have to do some validation
+    const userForToken = {
+        email: user.email,
+        id: user.id,
+    };
+
+    const token = jwt.sign(
```

Comment:
Do note that jwt.sign() will return a non-encrypted token, which can be pasted to eg. https://jwt.io to find out the contents. In this case, one can find out the email address inside the token. We can stick with this for now though.